### PR TITLE
Refactor `HashJoin` to avoid `too large translation units`

### PR DESCRIPTION
**Pull request by lgbo-ustc**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Instantiate template methods ahead in different .cpp files, avoid too large translation units during compiling.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All NOT Required Checks
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: Integration Tests
- [ ]  Exclude: Stateless tests
- [ ]  Exclude: Stateful tests
- [ ]  Exclude: Performance tests
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with Aarch64
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
---
- [ ]  Do not test
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit b8650152b04388094c2c62e20b638ccd8ce416c4 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64818/b8650152b04388094c2c62e20b638ccd8ce416c4/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by franz101** [comment]
To improve the Hash join it would be interesting to test clickhouse for the Hash join implementation:
https://cedardb.com/blog/simple_efficient_hash_tables/

**Comment by lgbo-ustc** [comment]
@vdimir  could you help to review this ?

File: src/Interpreters/HashJoin/HashJoin.h [hunk 1]
```diff
@@ -433,8 +403,7 @@ class HashJoin : public IJoin
     /// Number of this flags equals to hashtable buffer size (plus one for zero value).
     /// Changes in hash table broke correspondence,
     /// so we must guarantee constantness of hash table during HashJoin lifetime (using method setLock)
-    mutable JoinStuff::JoinUsedFlags used_flags;
-
+    mutable std::unique_ptr<JoinStuff::JoinUsedFlags> used_flags;
```
Comment (vdimir) [inline src/Interpreters/HashJoin/HashJoin.h:406]: Why is it changed to  unique_ptr?

**Review by vdimir — APPROVED** [review]

**Comment by vdimir** [comment]
Just in case simple script I used to check out unique added/removed lines 

<details>
<summary>Details</summary>

```python
#!/usr/bin/env python3

import argparse
import re
from collections import Counter
from fuzzywuzzy import fuzz

def parse_diff(diff_file_path):
    added_lines = []
    deleted_lines = []

    with open(diff_file_path, 'r') as file:
        for line in file:
            if line.startswith('+') and not line.startswith('+++'):
                added_lines.append(line[1:].strip())
            elif line.startswith('-') and not line.startswith('---'):
                deleted_lines.append(line[1:].strip())

    return added_lines, deleted_lines

def find_unique_changes(added_lines, deleted_lines):
    added_counter = Counter(added_lines)
    deleted_counter = Counter(deleted_lines)

    unique_added = [line for line in added_counter if added_counter[line] > 0 and deleted_counter[line] == 0]
    unique_deleted = [line for line in deleted_counter if deleted_counter[line] > 0 and added_counter[line] == 0]

    return unique_added, unique_deleted

def main():
    parser = argparse.ArgumentParser(description="Process a diff file to find unique added and deleted lines.")
    parser.add_argument('diff_file_path', type=str, help='Path to the diff file')

    args = parser.parse_args()
    diff_file_path = args.diff_file_path

    added_lines, deleted_lines = parse_diff(diff_file_path)
    unique_added, unique_deleted = find_unique_changes(added_lines, deleted_lines)

    def find_max_score(line, other_lines):
        max_score = 0
        max_score_line = ""
        for other_line in other_lines:
            score = fuzz.ratio(line, other_line)
            if score > max_score:
                max_score = score
                max_score_line = other_line
        if max_score > 90:
            return max_score_line
        return "NO MATCH"

    print(f"================== Unique Added Lines ({len(unique_added)}):")
    for i, line in enumerate(unique_added, 1):
        max_score_line = find_max_score(line, unique_deleted)
        print(f"\t{i}. {line}\n\t\t{max_score_line}")


    print()
    print(f"================== Unique Deleted Lines ({len(unique_deleted)}):")
    for i, line in enumerate(unique_deleted, 1):
        max_score_line = find_max_score(line, unique_added)
        print(f"\t{i}. {line}\n\t\t{max_score_line}")



if __name__ == "__main__":
    main()

```

</details>

Output (for https://patch-diff.githubusercontent.com/raw/ClickHouse/ClickHouse/pull/64818.diff):

https://pastila.nl/?001ab97c/1ddef49844bbaf80f23fe2c35c578adc#qGUMTIAqnHG9IVQU0Go8lw==

File: src/Interpreters/HashJoin/HashJoin.h (hunk 1 above)
Comment (lgbo-ustc) [reply src/Interpreters/HashJoin/HashJoin.h:406]: to avoid include the header file which would cause nested including

**Pull request merged**